velox/functions/sparksql/specialforms/AtLeastNNonNulls.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/AtLeastNNonNulls.h"
+#include "velox/expression/SpecialForm.h"
+
+using namespace facebook::velox::exec;
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+class AtLeastNNonNullsExpr : public SpecialForm {
+ public:
+  AtLeastNNonNullsExpr(
+      TypePtr type,
+      std::vector<ExprPtr>&& inputs,
+      bool inputsSupportFlatNoNullsFastPath)
+      : SpecialForm(
+            std::move(type),
+            std::move(inputs),
+            AtLeastNNonNullsCallToSpecialForm::kAtLeastNNonNulls,
+            inputsSupportFlatNoNullsFastPath,
+            false /* trackCpuUsage */) {}
+
+  void evalSpecialForm(
+      const SelectivityVector& rows,
+      EvalCtx& context,
+      VectorPtr& result) override {
+    context.ensureWritable(rows, type(), result);
+    (*result).clearNulls(rows);
+    auto flatResult = result->asFlatVector<bool>();
+    LocalSelectivityVector activeRowsHolder(context, rows);
+    auto activeRows = activeRowsHolder.get();
+    VELOX_DCHECK_NOT_NULL(activeRows);
+    // Load 'n' from the first input.
+    VectorPtr vector;
+    inputs_[0]->eval(*activeRows, context, vector);
+    VELOX_CHECK(!context.errors());
+    VELOX_USER_CHECK(
+        vector->isConstantEncoding(),
+        "The first input should be constant encoding.");
+    auto constVector = vector->asUnchecked<ConstantVector<int32_t>>();
+    VELOX_USER_CHECK(
+        !constVector->isNullAt(0), "The first parameter should not be null.");
+    const int32_t n = constVector->valueAt(0);
```

Comment:
It looks like we can check for constant expression and extract the constant value in the 'constructSpecialForm' method. In this way, we don't need to extract it in each eval. Some example:

https://github.com/facebookincubator/velox/blob/f1d9502e4c6d0a9b8507fb6b8aa0eb5befe40ff4/velox/functions/sparksql/specialforms/DecimalRound.cpp#L236

Reply:
We only have the ExprPtr, so we need the `EvalCtx` to get the value from it.  `EvalCtx` is not available in the function `constructSpecialForm`. Please let me know if I miss anything here. Thanks.

Reply:
We can cast it to ConstantExpr, and value() returns a  constant vector.

https://github.com/facebookincubator/velox/blob/f1d9502e4c6d0a9b8507fb6b8aa0eb5befe40ff4/velox/functions/sparksql/specialforms/DecimalRound.cpp#L231-L232

Reply:
I see, updated, thanks.